test/framework/src/main/java/org/elasticsearch/transport/nio/channel/AbstractNioChannel.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.transport.nio.ESSelector;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.NetworkChannel;
+import java.nio.channels.SelectableChannel;
+import java.nio.channels.SelectionKey;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public abstract class AbstractNioChannel<S extends SelectableChannel & NetworkChannel> implements NioChannel {
```

Comment:
At times java generics confuse me, but I do not think you can do this:

`TcpTransport<AbstractNioChannel<S extends SelectableChannel & NetworkChannel>>`

Which means that instead you have to do this:

`TcpTransport<AbstractNioChannel>`

And then you get a bunch of warnings when compiling and stuff. That is why I introduced the interface.

Let me know if you have a solution that can drop the interface and work well with TCPTransport that I am not thinking of.